include/soft_limiter.h

```diff
@@ -0,0 +1,255 @@
+	constexpr static float upper_bound = static_cast<float>(out_limits::max());
+	constexpr static float lower_bound = static_cast<float>(out_limits::min());
+
+	bool FindPeaks(const in_array_t &stream, uint16_t frames) noexcept;
+	void Release() noexcept;
+
+	std::string channel_name = {};
+	AudioFrame limit_scale = {1, 1}; // real-time limit applied to stream
+	AudioFrame peak = {1, 1};        // holds real-time peak amplitudes
+	const AudioFrame &prescale;      // scale before operating on the stream
+	int limited_ms = 0;              // milliseconds that needed limiting
+	int non_limited_ms = 0; // milliseconds that didn't need limiting
+};
+
+template <size_t array_frames>
+SoftLimiter<array_frames>::SoftLimiter(const std::string &name, const AudioFrame &scale)
+        : channel_name(name),
+          prescale(scale)
+{
+	static_assert(array_frames > 0, "need some quantity of frames");
+	static_assert(array_frames < 16384, "too many frames adds audible latency");
+}
+
+template <size_t array_frames>
+bool SoftLimiter<array_frames>::FindPeaks(const in_array_t &stream,
+                                          const uint16_t samples) noexcept
+{
+	auto val = stream.begin();
+	const auto val_end = stream.begin() + samples;
+	while (val < val_end) {
+		// Left channel
+		const auto amplitude_left = fabsf(*val) * prescale.left;
+		peak.left = std::max(peak.left, amplitude_left);
+		++val;
+
+		// Right channel
+		const auto amplitude_right = fabsf(*val) * prescale.right;
+		peak.right = std::max(peak.right, amplitude_right);
+		++val;
+	}
+	const bool limiting_needed = (peak.left > upper_bound ||
+	                              peak.right > upper_bound);
+
+	// Calculate the percent we need to scale down each channel. In cases
+	// where one channel is less than the upper-bound, its time_ratio is limited
+	// to 1.0 to retain the original level.
+	if (limiting_needed)
+		limit_scale = {std::min(1.0f, upper_bound / peak.left),
+		               std::min(1.0f, upper_bound / peak.right)};
+
+	return limiting_needed;
+	// LOG_MSG("%s peak left = %f", channel_name.c_str(),
+	// static_cast<double>(peak.left));
+}
+
+template <size_t array_frames>
+void SoftLimiter<array_frames>::Apply(const in_array_t &in,
+                                      out_array_t &out,
+                                      const uint16_t req_frames) noexcept
+{
+	// Ensure the buffers are large enough to handle the request
+	const uint16_t samples = req_frames * 2; // left and right channels
+	assert(samples <= in.size());
+
+	const bool limiting_needed = FindPeaks(in, samples);
+
+	// get our in and out iterators
+	auto in_val = in.begin();
+	const auto in_val_end = in.begin() + samples;
+	auto out_val = out.begin();
+
+	// Process the signal by pre-scaling and limit-scaling
+	// Note: if limit-scaling isn't needed then its values are simply 1.0
+	while (in_val < in_val_end) {
+		*out_val++ = static_cast<int16_t>(*in_val++ * prescale.left *
```

Comment:
Optimization: lift the prescale * limit_scale calc above the loop, and store the two as consts (compiler likely does this, but it's intuitive and might improve readability).